src/middlewared/middlewared/plugins/pool.py

```diff
@@ -2230,15 +2220,34 @@ def unlock(self, job, id, options):
 
             try:
                 self.middleware.call_sync(
-                    'zfs.dataset.load_key', name, {'key': datasets[name]['key'], 'mount': True, 'recursive': True}
+                    'zfs.dataset.load_key', name, {'key': datasets[name]['key'], 'mount': False}
                 )
             except CallError as e:
                 failed[name]['error'] = 'Invalid Key' if 'incorrect key provided' in str(e).lower() else str(e)
             else:
-                unlocked.append(name)
+                # Before we mount the dataset in question, we should ensure that the path where it will be mounted
+                # is not already being used by some other service/share. In this case, we should simply rename the
+                # directory where it will be mounted
+
+                mount_path = os.path.join('/mnt', name)
+                if os.path.exists(mount_path):
+                    if not os.path.isdir(mount_path) or os.listdir(mount_path):
```

Comment:
We should exclude `.zfs` directory when checking if dataset is empty or not (in case of `snapdir=visible`). We do this in other modules too, let's create an util function for that.

Reply:
At this stage the dataset is not mounted, so `.zfs` will not exist. We can create a util function but that is not useful for this purpose ? It's just checking if the path exists and if it's empty or not. When we umount a dataset ( which is done first when a dataset is locked ), the path still exists.

Reply:
My bad, you are right